core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java

```diff
@@ -0,0 +1,104 @@
+package org.activityinfo.model.type.subform;
+/*
+ * #%L
+ * ActivityInfo Server
+ * %%
+ * Copyright (C) 2009 - 2013 UNICEF
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+import com.google.common.collect.Maps;
+import org.activityinfo.model.form.FormClass;
+import org.activityinfo.model.legacy.CuidAdapter;
+import org.activityinfo.model.resource.ResourceId;
+
+import java.util.Map;
+
+/**
+ * Represents type of form classes : locationType classes or activity classes.
+ * Mainly used to allow user define subform kind.
+ *
+ * @author yuriyz on 03/12/2015.
+ */
+public enum ClassType {
```

Comment:
It is actually already implemented in this way, maybe a bit different. `ClassType` is just convenient way to bind legacy model and new model. It is super-type, ClassType.partner means fetch all partners for given db. Here are a bit more details:

Reference constants : https://github.com/bedatadriven/activityinfo/blob/ai-664/core/model/src/main/java/org/activityinfo/model/type/subform/SubformConstants.java

Subform is described by `FormClass`. Now I'm storing in subform `FormClass` with two special form fields:

1. Reference on value type `FormClass`.

```
 ReferenceType typeClass = (ReferenceType) subForm.getField(SubformConstants.TYPE_FIELD_ID).getType();
 ResourceId typeClassId = typeClass.getRange().iterator().next();
```

Under value type `FormClass` I mean form class for which we will fetch/store instances. Like Partner, Location Type, Period or user-defined form class. Theoretically it can be
any `FormClass` which we support. 

Here is translation from legacy to new model in QueryExecutor (internally check is made by pre-defined ids `hardcoded` in ClassType.).
```
 private Promise<List<FormInstance>> instancesByClassType(ParentCriteria.Parent parent) {
        ResourceId parentId = parent.getParentId();

        if (parentId.equals(ClassType.LOCATION_TYPE.getResourceId())) {
            return locationTypes(parent);
        } else if (parentId.equals(ClassType.PROJECT.getResourceId())) {
            return projects(parent);
        } else if (parentId.equals(ClassType.PARTNER.getResourceId())) {
            return partner(parent);
        } else {
            throw new UnsupportedOperationException("ClassType is not supported, classType: " + parentId);
        }
    }
```

2. Metadata formfield. For example we want to have different tab numbers for specific subform. E.g. Location Type may have long names and we may want to keep 4 tabs, while 
for period it may be 10. 

```
QuantityType tabsCountType = (QuantityType) subForm.getField(SubformConstants.TAB_COUNT_FIELD_ID).getType();
```

In this way we can store in subform `FormClass` any number of metadata we need (for now it is only tab count and value type `FormClass`).

To summarize, current model is generic and we may support any value type `FormClass`.